### PR TITLE
fix(sitemap): exclude pages with robots noindex from sitemap

### LINKED ISSUE #6247: Pages with noindex meta still showing in sitemap
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#reporting-new-issues).

### Prerequisites

- [X] I'm using the latest version of Docusaurus.
- [ ] I have tried the `npm run clear` or `yarn clear` command.
- [ ] I have tried `rm -rf node_modules yarn.lock package-lock.json` and re-installing packages.
- [ ] I have tried creating a repro with https://new.docusaurus.io.
- [ ] I have read the console error message carefully (if applicable).

### Description

It is bad practice to add noindex pages to sitemap:

- https://sitebulb.com/hints/xml-sitemaps/noindex-url-in-xml-sitemaps/#:~:text=If%20a%20URL%20is%20noindex,in%20unintended%20URLs%20getting%20indexed
- https://support.google.com/webmasters/thread/78460212/is-meta-robots-noindex-for-page-listed-in-sitemap-xml-ok?hl=en

### Steps to reproduce

1. Add `<meta name="robots" content="noindex">` to the head tag of a page.
2. Build.

### Expected behavior

The sitemap doesn't include that page.

### Actual behavior

It does—we don't filter routes by `noindex`.

We should definitely conform to the `noIndex` config option and not output a sitemap in that case. For individual pages, we should probably read the HTML files and filter those with `noindex`? Or should we ask the user to provide a list of routes to ignore? (I think we should have both)

### Your environment

_No response_

### Reproducible demo

_No response_

### Self-service

- [X] I'd be willing to fix this bug myself.

### PR DESCRIPTION
**Pull request by Josh-Cena**
## Motivation

fix #6247. I'm not sure if providing the entire helmet state would be too big as part of the plugin API, but currently it's the only way to get structured data about the head state. Otherwise, we can only export a string and then parse it again in `postBuild`, which sounds like too much overhead.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

I reverted #7122 and added `<meta name="robots" content="noindex" />` to the debug layout. It turns out that these routes are still not included in the sitemap.

### CODE REVIEW

File: packages/docusaurus-plugin-sitemap/src/createSitemap.ts [hunk 1]
```diff
@@ -10,10 +10,13 @@ import type {PluginOptions} from '@docusaurus/plugin-sitemap';
 import type {DocusaurusConfig} from '@docusaurus/types';
 import {applyTrailingSlash} from '@docusaurus/utils-common';
 import {createMatcher} from '@docusaurus/utils';
+import type {HelmetServerState} from 'react-helmet-async';
+import type {ReactElement} from 'react';
 
 export default async function createSitemap(
   siteConfig: DocusaurusConfig,
   routesPaths: string[],
+  helmet: {[location: string]: HelmetServerState},
```
Comment (slorber) [inline packages/docusaurus-plugin-sitemap/src/createSitemap.ts:19]: maybe rename to head?

Comment (slorber) [inline packages/docusaurus-plugin-sitemap/src/createSitemap.ts:19]: maybe rename to head?

File: packages/docusaurus/src/commands/build.ts [hunk 2]
```diff
@@ -224,7 +229,11 @@ async function buildLocale({
       if (!plugin.postBuild) {
         return;
       }
-      await plugin.postBuild({...props, content: plugin.content});
+      await plugin.postBuild({
+        ...props,
+        helmet: headTags,
```
Comment (slorber) [inline packages/docusaurus/src/commands/build.ts:234]: maybe rename to head or headTags?

Not fan of using "helmet" name here, lib is an implementation detail and we already use "head" name in other places

Also not fan of the API surface it introduces.

Users shouldn't use something like `helmet[route]?.meta.toComponent()`, this is a quite confusing that we'd struggle to document well and ensure retrocompatibility on.

Maybe we could only expose a few route attributes and transform all `HelmetDatum` types to a convenient format, easy to document?

Comment (slorber) [inline packages/docusaurus/src/commands/build.ts:234]: maybe rename to head or headTags?

Not fan of using "helmet" name here, lib is an implementation detail and we already use "head" name in other places

Also not fan of the API surface it introduces.

Users shouldn't use something like `helmet[route]?.meta.toComponent()`, this is a quite confusing that we'd struggle to document well and ensure retrocompatibility on.

Maybe we could only expose a few route attributes and transform all `HelmetDatum` types to a convenient format, easy to document?

**Review by slorber — CHANGES_REQUESTED** [review]
Thanks 👍 

That looks to work nicely

Not fan of "helmet" name + public API surface, which makes it complex to document (also doc is missing)

File: packages/docusaurus-plugin-debug/src/index.ts [hunk 3]
```diff
@@ -9,8 +9,6 @@ import type {LoadContext, Plugin} from '@docusaurus/types';
 import {docuHash, normalizeUrl, posixPath} from '@docusaurus/utils';
 import path from 'path';
 
-export const routeBasePath = '__docusaurus/debug';
```
[the remark this answers is not in the document]
Comment (Josh-Cena) [reply packages/docusaurus-plugin-debug/src/index.ts:12]: At least not as a public API. I also don't find this abstraction particularly useful

**Comment by slorber** [comment]
> Not a fan of the API design either, but I can't find a nicer abstraction of the helmet state in a structured way.

I suggest we move on and merge this as it's a nice improvement for internal usage in the sitemaps plugin

We can document this behind an experimental flag, warn it may change and recommend to use "head" only in site plugins, not public/open-source plugins?

We don't flag any feature as experimental for now but we probably could as we have a few undocumented apis (like `useRouteContext`)

**Comment by Josh-Cena** [comment]
We also have a `routes` prop passed to `postBuild` that's undocumented. Maybe we'll eventually remove that

**Comment by slorber** [comment]
> We also have a routes prop passed to postBuild that's undocumented. Maybe we'll eventually remove that

🤷‍♂️ we might as well keep `routes` and assign the rendered metadata to items instead of having a separate object with head metadata. Both work, I just don't know what is more convenient to use (considering routes are deeply nested, metadata would only be added to the leaves final routes)

**Comment by slorber** [comment]
I guess we can keep all this undocumented for now, let's merge as is 👍 

Hopefull if someone plans to use this they'll reach out and explain their use-case

**Review by slorber — APPROVED** [review]

**Pull request merged**